### ISSUE TITLE
Compaction after writing shorter streams.

### ISSUE THREAD

**Issue by francisdb** [mdsteele/rust-cfb#55]
I have the impression that the files never shrink after reducing some streams in size.
Have not tested removing streams but could also be affected?

Is there a function available to force a compaction?

Maybe a compact method that looks like the copy code shown in #33?

**Comment by McSpidey** [comment]
I’m curious about this too. Have you run any tests to confirm it?

**Comment by mdsteele** [comment]
Correct, the `cfb` crate currently never shrinks the underlying file.  The basic problem here is that `cfb::CompoundFile` is generic over the underlying reader/writer type (it could be a `File`, or a `Cursor`, or a `cfb::Stream`, or something else entirely), but as far as I know there's no standard Rust trait that provides a generic way to truncate/resize a file-like object.  Possibly the best solution is for the `cfb` crate to provide such a trait, along with implementations for `File` and other `std` types.

**Comment by mdsteele** [comment]
See https://github.com/rust-lang/rfcs/issues/1376

**Comment by McSpidey** [comment]
Just logging I’ve been doing some basic I/O testing with a simple text file and it’s looking like std::fs::File::set_len (https://doc.rust-lang.org/std/fs/struct.File.html#method.set_len) with seek rewind for the cursor (https://doc.rust-lang.org/std/io/trait.Seek.html#method.rewind) can work to shrink a file. There’s also this truncate in OpenOptons (https://doc.rust-lang.org/std/fs/struct.OpenOptions.html#method.truncate) but it looks like it does something different.

**Comment by McSpidey** [comment]
I've been doing some reading on what MS considers standard process for compacting a CFB and it appears they want you use copyto  to make a new file anyway, then stream the new compacted data into it rather than overwrite in place.
https://learn.microsoft.com/en-us/windows/win32/api/objidl/nf-objidl-istorage-copyto
"Note  To compact a document file, call CopyTo on the root storage object and copy to a new storage object."

Even this doc says you should use copyto for save-as.
https://learn.microsoft.com/en-us/windows/win32/stg/irootstorage-compound-file-implementation
"applications performing a save-as operation should manually create the second compound file and invoke [IStorage::CopyTo](https://learn.microsoft.com/en-us/windows/desktop/api/Objidl/nf-objidl-istorage-copyto)."

This makes things way simpler for the library. Apps that want to overwrite in place can then do an atomic file replace with the new compacted copy (combined delete+rename in the kernel)
